File: backend/services/citations_service.py
import re
from typing import Dict, Any, Optional
from backend.database import STANDARDS_DB

OFFICIAL_BIS_PORTAL_BASE = "https://www.services.bis.gov.in/php/BIS_2.0/bisconnect/knowyourstandards/"

class CitationsService:
    @staticmethod
    def get_citation_by_id(citation_id: str) -> Optional[Dict[str, Any]]:
        clean_id = citation_id.lower().strip()

        for std in STANDARDS_DB:
            for clause in std.get("clauses", []):
                cl_slug = re.sub(r"[^a-z0-9]", "", clause.get("number", "").lower())
                generated_id = f"{std['id']}-{cl_slug}"

                if generated_id == clean_id or clause.get("id", "").lower() == clean_id:
                    return {
                        "citationId": generated_id,
                        "standardId": std["id"],
                        "standardCode": std["code"],
                        "standardTitle": std["title"],
                        "clauseNumber": clause["number"],
                        "clauseTitle": clause["title"],
                        "textSnippet": clause["content"],
                        "testRequirement": clause.get("testRequirement") or clause.get("testMethod"),
                        "division": std.get("division") or std.get("department", ""),
                        "officialBisPortalUrl": OFFICIAL_BIS_PORTAL_BASE
                    }

        return None

File: src/server/services/citations.service.ts
import { STANDARDS_DATABASE } from "@/lib/standards-data";
import { NotFoundError } from "../utils/errors";

export interface SourceCitationDetail {
  citationId: string;
  standardId: string;
  standardCode: string;
  standardTitle: string;
  clauseNumber: string;
  clauseTitle: string;
  textSnippet: string;
  testRequirement?: string;
  division: string;
  officialBisPortalUrl: string;
}

export class CitationsService {
  static getCitationById(id: string): SourceCitationDetail {
    // Expected id format: "std_id-clause_num" e.g. "is-17526-2021-5.1" or "is-1293-2019-cl6"
    for (const std of STANDARDS_DATABASE) {
      for (const clause of std.clauses) {
        const generatedId = `${std.id}-${clause.number.toLowerCase().replace(/[^a-z0-9]/g, "")}`;
        if (generatedId === id.toLowerCase() || clause.id === id) {
          return {
            citationId: generatedId,
            standardId: std.id,
            standardCode: std.code,
            standardTitle: std.title,
            clauseNumber: clause.number,
            clauseTitle: clause.title,
            textSnippet: clause.content,
            testRequirement: clause.testRequirement || clause.testMethod,
            division: std.division || std.department,
            officialBisPortalUrl: "https://www.services.bis.gov.in/php/BIS_2.0/bisconnect/knowyourstandards/",
          };
        }
      }
    }

    throw new NotFoundError(`Citation with ID '${id}'`);
  }
}
